s3/s3_operations.cpp

```diff
@@ -845,6 +849,55 @@ namespace irods_s3 {
 
             irods::error result = SUCCESS();
 
+            irods::file_object_ptr file_obj = boost::dynamic_pointer_cast<irods::file_object>(_ctx.fco());
+
+            // create entry for fd in per_thread_data if it doesn't exist
+            int fd = file_obj->file_descriptor();
+            per_thread_data data{};
+            if (fd_data.exists(fd)) {
+                data = fd_data.get(fd);
+            }
+
+            // Issue 2124 - On a PUT operation with checksum, the open_mode was forced to O_CREAT, O_WRONLY, O_TRUNC so that streaming
+            // would work.  We need to check if this was a PUT_OPR and if so we have moved on to the read-only portion of
+            // processing for checksum calculation.  If that is the case, the open_mode will be reset to O_RDONLY.
+            // So that this isn't done on every read call, a flag (checked_for_checksum_read_after_put) is set to indicate that we
+            // have already checked for this condition and adjusted the open_mode accordingly.
+            if (!data.checked_for_checksum_read_after_put) {
+
+                // Get oprType
+                int oprType = -1;
+                bool found = false;
+                for (int i = 0; i < NUM_L1_DESC; ++i) {
```

Comment:
Consider using a `std::span` here as a "view" of the L1 descriptor table. This would also allow you to use a range-based for loop.

Feel free to ignore this as what you're doing here is standard practice (for now)

Reply:
Sadly, we can't use `std::span` here. It's a C++20 feature.

Reply:
D'oh. 4-2-stable